File: src/cpp/cdm/system/equipment/mechanical_ventilator/actions/SEMechanicalVentilatorHold.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
#include "cdm/CommonDefs.h"
#include "cdm/system/equipment/mechanical_ventilator/actions/SEMechanicalVentilatorHold.h"
#include "cdm/system/equipment/mechanical_ventilator/SEMechanicalVentilator.h"
#include "cdm/substance/SESubstance.h"
#include "cdm/substance/SESubstanceManager.h"
#include "cdm/substance/SESubstanceConcentration.h"
#include "cdm/substance/SESubstanceFraction.h"
#include "cdm/properties/SEScalar0To1.h"
#include "cdm/properties/SEScalarMassPerVolume.h"
#include "cdm/properties/SEScalarPressure.h"
#include "cdm/properties/SEScalarPressureTimePerVolume.h"
#include "cdm/properties/SEScalarVolumePerTime.h"
#include "cdm/properties/SEScalarVolume.h"
#include "cdm/properties/SEScalarTime.h"
#include "cdm/io/protobuf/PBEquipmentActions.h"

SEMechanicalVentilatorHold::SEMechanicalVentilatorHold(Logger* logger) : SEMechanicalVentilatorAction(logger)
{
  m_AppliedRespiratoryCycle = eAppliedRespiratoryCycle::Expiratory;
  m_State = eSwitch::Off;
}

SEMechanicalVentilatorHold::~SEMechanicalVentilatorHold()
{

}

void SEMechanicalVentilatorHold::Clear()
{
  SEMechanicalVentilatorAction::Clear();
  m_AppliedRespiratoryCycle = eAppliedRespiratoryCycle::Expiratory;
  m_State = eSwitch::Off;
}

void SEMechanicalVentilatorHold::Copy(const SEMechanicalVentilatorHold& src, bool /*preserveState*/)
{// Using Bindings to make a copy
  PBEquipmentAction::Copy(src, *this);
}

bool SEMechanicalVentilatorHold::IsValid() const
{
  return SEMechanicalVentilatorAction::IsValid();
}

bool SEMechanicalVentilatorHold::IsActive() const
{
  if (!SEMechanicalVentilatorAction::IsActive())
    return false;
  return GetState() == eSwitch::On;
}
void SEMechanicalVentilatorHold::Deactivate()
{
  SEMechanicalVentilatorAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SEMechanicalVentilatorHold::GetScalar(const std::string& /*name*/)
{
  return nullptr;
}

void SEMechanicalVentilatorHold::SetAppliedRespiratoryCycle(eAppliedRespiratoryCycle c)
{
  m_AppliedRespiratoryCycle = c;
}
eAppliedRespiratoryCycle SEMechanicalVentilatorHold::GetAppliedRespiratoryCycle() const
{
  return m_AppliedRespiratoryCycle;
}

void SEMechanicalVentilatorHold::SetState(eSwitch s)
{
  m_State = s;
}
eSwitch SEMechanicalVentilatorHold::GetState() const
{
  return m_State;
}

void SEMechanicalVentilatorHold::ToString(std::ostream &str) const
{
  SEMechanicalVentilatorAction::ToString(str);
  str << "\n\tAppliedRespiratoryCycle: " << eAppliedRespiratoryCycle_Name(m_AppliedRespiratoryCycle);
  str << "\n\tState: " << eSwitch_Name(m_State);
  str << std::flush;
}

File: src/csharp/pulse/cdm/system/equipment/mechanical_ventilator/actions/SEMechanicalVentilatorHold.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SEMechanicalVentilatorHold : SEMechanicalVentilatorAction
  {
    protected eSwitch state;
    protected eAppliedRespiratoryCycle appliedRespiratoryCycle;

    public SEMechanicalVentilatorHold()
    {
      Clear();
    }

    public SEMechanicalVentilatorHold(SEMechanicalVentilatorHold other)
    {
      Copy(other);
    }

    public void copy(SEMechanicalVentilatorHold other)
    {
      base.Copy(other);
      state = other.state;
      appliedRespiratoryCycle = other.appliedRespiratoryCycle;
    }

    public override void Clear()
    {
      base.Clear();
      appliedRespiratoryCycle = eAppliedRespiratoryCycle.Expiratory;
      state = eSwitch.Off;
    }

    public override bool IsValid()
    {
      return true;
    }

    public eSwitch GetState()
    {
      return state;
    }
    public void SetState(eSwitch s)
    {
      state = s;
    }

    public eAppliedRespiratoryCycle GetAppliedRespiratoryCycle()
    {
      return appliedRespiratoryCycle;
    }
    public void SetAppliedRespiratoryCycle(eAppliedRespiratoryCycle c)
    {
      appliedRespiratoryCycle = c;
    }

    public override string ToString()
    {
      string str = "Mechanical Ventilator Hold";
      str += "\n\tState: " + this.state;
      str += "\n\tApplied Respiratory Cycle: " + this.appliedRespiratoryCycle;

      return str;
    }
  }
}
